src/tools/wasm-fuzz-types.cpp

```diff
@@ -561,7 +501,7 @@ void Fuzzer::checkInhabitable() {
   std::unordered_set<HeapType> visited, visiting;
   bool haveUninhabitable = false;
   for (auto type : types) {
-    if (findUninhabitable(type, visited, visiting)) {
+    if (HeapTypeGenerator::findUninhabitable(type, visited, visiting)) {
```

Comment:
To make `findUninhabitable` an implementation detail, we can use `haveUninhabitable = HeapTypeGenerator::getUninhabitable(types).size() == types.size())`. That will let us make `findUninhabitable` return bool and simplify the implementation as well.

Reply:
Hmm, there is code lower down that looks at individual types too, though? That's why atm `findUninhabitable` is not just an implementation detail. But that later code could be removed though if we think those checks + fatal errors are not needed, like this perhaps?
```diff
diff --git a/src/tools/wasm-fuzz-types.cpp b/src/tools/wasm-fuzz-types.cpp
index d10510222..1f0aa1b4a 100644
--- a/src/tools/wasm-fuzz-types.cpp
+++ b/src/tools/wasm-fuzz-types.cpp
@@ -498,27 +498,9 @@ void Fuzzer::checkInhabitable() {
   }
 
   // Check whether any of the original types are uninhabitable.
-  std::unordered_set<HeapType> visited, visiting;
-  bool haveUninhabitable = false;
-  for (auto type : types) {
-    if (HeapTypeGenerator::findUninhabitable(type, visited, visiting)) {
-      haveUninhabitable = true;
-      break;
-    }
-  }
-  visited.clear();
-  visiting.clear();
-
-  if (haveUninhabitable) {
-    // Verify that the transformed types are inhabitable.
-    for (auto type : inhabitable) {
-      if (auto uninhabitable =
-            HeapTypeGenerator::findUninhabitable(type, visited, visiting)) {
-        IndexedTypeNameGenerator print(inhabitable);
-        Fatal() << "Found uninhabitable type: " << print(*uninhabitable);
-      }
-    }
-  } else if (getTypeSystem() == TypeSystem::Isorecursive) {
+  haveUninhabitable = HeapTypeGenerator::getInhabitable(types).size() != types.size();
+
+  if (!haveUninhabitable && getTypeSystem() == TypeSystem::Isorecursive) {
     // Verify the produced inhabitable types are the same as the original types.
     if (types.size() != inhabitable.size()) {
       Fatal() << "Number of inhabitable types does not match number of "
```
I'd also be ok leaving this for after this lands as you said.

Reply:
The way I would fix this would be to make that "Found uninhabitable type" error actually print all of the uninhabitable types rather than the arbitrary uninhabitable type we happen to find first. We could do `set(types) \ `set(inhabitable)` types to find the uninhabitable types.

Reply:
But the set `inhabitable` would be generated by filtering out the uninhabitable ones. So we'd be using the exact same process twice to filter to inhabitable types? Given that I think we can just delete it here. But I feel like I might be missing something.

Reply:
no, I'd use a normal membership test on a `std::unordered_set` of the uninhabitable types to find inhabitable types. It's ok to do some extra slow work in this failure case to print a more useful message.

Reply:
Ok, I tried to write it that way, PTAL.

I suspect I have done something wrong, as this now fails your fuzzer,
```
$ ./bin/wasm-fuzz-types
Iteration 1
Running with seed 5435209648995731794
Fatal: makeInhabitable incorrectly changed type (type $0 (array i16))
```
Maybe I don't quite understand how to use your APIs properly? Could be something obvious in the last commit that I'm doing wrong that you can see.

Reply:
(It seems line new line 528 might be wrong, as it should only look at uninhabitable types..? But that line is unchanged in this PR, so I'm confused.)

Reply:
(But to look at inhabitable types specifically we'd need that helper function `findUninhabitable` again... which we've said we want to avoid here.)

Reply:
Oh, I just realized I wrote `set(inhabitable)` above when I meant `set(uninhabitable)`. Sorry for the confusion! Taking a look at the latest code now...